### PR TITLE
Callout microapp enhancements

### PR DESCRIPTION
**Pull request by shubham7109**
Related to issue: `#4116`

### Description:
PR to enhance and update the Callout MicroApp FeatureScreen and TapLocationScreen. 

### Summary of changes:

- TapLocationScreen
    - Updated screen to use a fab to show CalloutOptions in a bottom sheet
    - Added an "X" close button on the Callout to clear. 
    - Added HtmlText to show static HTML content on the Callout
    - MapView uses device width for insets, resulting in the Callout content always being half the device screen width. (For, testing insets functionality in any orientation) 
- Feature Screen
    - Updated screen to use a fab to show CalloutOptions in a bottom sheet
    - Added an "X" close button on the Callout to clear.
    - Updated the [WebMap](https://www.arcgis.com/home/item.html?id=16f1b8ba37b44dc3884afc8d5f454dd2), which contains Polygons (neighborhood) and Point (trees) features.
    - Added `recenterMap` function to recenter map on tapped point. (Helpful for testing, passing null tapped location)
  
### Pre-merge Checklist

- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/) Job for this PR has been run
  - [x] link: ⛔  vTests are broken. 
- Unit and/or integration tests have been added to exercise this PR's logic, and the tests are passing:
  - [ ] Yes
  - [x] No

### CODE REVIEW

**Comment by shubham7109** [comment]
**Note:** I will rebase this PR once #466 & #462 is merged.

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/FeatureScreen.kt [hunk 1]
```diff
@@ -18,26 +18,229 @@
...
+ */
 
+@OptIn(ExperimentalMaterial3Api::class)
 @Composable
-fun FeatureScreen(){
-    Box{
+fun FeatureScreen(viewModel: MapViewModel) {
+    val selectedGeoElement = viewModel.selectedGeoElement.collectAsState().value
+    var calloutVisibility by rememberSaveable { mutableStateOf(true) }
+    var nullTapLocation by rememberSaveable { mutableStateOf(false) }
```
Comment (shubham7109) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/FeatureScreen.kt:74]: > https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/pull/459#discussion_r1640342227

### To Address: 
Update this to `specifyTapLocation` / `useTapLocation`?

**Comment by shubham7109** [comment]
@puneet-pdx, ready for your review.

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt [hunk 2]
```diff
@@ -51,14 +51,24 @@ class MapViewModel : ViewModel() {
...
+        uri = "https://www.arcgis.com/home/item.html?id=16f1b8ba37b44dc3884afc8d5f454dd2"
+    ).apply {
+        initialViewpoint = Viewpoint(
+            Point(x = -1.3659e7, y = 5.6917e6),
+            scale = 50000.0,
+        )
+    }
 
     private val _mapPoint = MutableStateFlow<Point?>(null)
     val mapPoint: StateFlow<Point?> = _mapPoint
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt:64]:
```suggestion
    val mapPoint: StateFlow<Point?> = _mapPoint.asStateFlow()
```

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt [hunk 3]
```diff
@@ -51,14 +51,24 @@ class MapViewModel : ViewModel() {
...
 
     private val _selectedGeoElement = MutableStateFlow<GeoElement?>(null)
     val selectedGeoElement: StateFlow<GeoElement?> = _selectedGeoElement
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt:67]:
```suggestion
    val selectedGeoElement: StateFlow<GeoElement?> = _selectedGeoElement.asStateFlow()
```

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt [hunk 4]
```diff
@@ -51,14 +51,24 @@ class MapViewModel : ViewModel() {
...
 
+    private val _selectedLayerName = MutableStateFlow("")
+    val selectedLayerName: StateFlow<String> = _selectedLayerName
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt:70]:
```suggestion
    val selectedLayerName: StateFlow<String> = _selectedLayerName.asStateFlow()
```

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt [hunk 5]
```diff
@@ -51,14 +51,24 @@ class MapViewModel : ViewModel() {
...
+
     private val _tapLocation = MutableStateFlow<Point?>(null)
     val tapLocation: StateFlow<Point?> = _tapLocation
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/MapViewModel.kt:73]:
```suggestion
    val tapLocation: StateFlow<Point?> = _tapLocation.asStateFlow()
```

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt [hunk 6]
```diff
@@ -67,59 +78,118 @@ fun TapLocationScreen(viewModel: MapViewModel) {
 
     var calloutVisibility by rememberSaveable { mutableStateOf(true) }
     var rotateOffsetWithGeoView by rememberSaveable { mutableStateOf(false) }
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt:80]: moving this to the MapViewModel causes the desired recomposition to occur when the state changes, which doesn't lead to the disappearing of the callout.

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt [hunk 7]
```diff
@@ -67,59 +78,118 @@ fun TapLocationScreen(viewModel: MapViewModel) {
...
+                                HtmlText(
+                                    html = "<b>Tapped location</b>:<br>" +
+                                            "<i>x</i>    = ${mapPoint.x.roundToInt()}<br>" +
+                                            "<i>y</i>    = ${mapPoint.y.roundToInt()}<br>" +
+                                            "<i>wkid</i> = ${mapPoint.spatialReference?.wkid}",
+                                    htmlFlag = HtmlCompat.FROM_HTML_MODE_COMPACT
+                                )
+                            }
+
+                            IconButton(onClick = viewModel::clearMapPoint) {
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt:133]: I am wondering if we really need this Icon as the toggle for dismissing the callout is present in the bottomsheet menu

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/FeatureScreen.kt [hunk 8]
```diff
@@ -19,116 +19,235 @@
...
-
-// TODO Case b.
-//- Show a MapView with a map with a Feature layer with features
-//- Display callout using Point(tap location) on a feature (with some text)
-//- Display a graphic at the tapped location
-//- add switch to enable/disable animation
 
+/**
+ * Displays a composable [MapView] to show a Callout on the tapped Feature.
+ */
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/FeatureScreen.kt:66]: I am seeing undesired behavior with this screen. Let's discuss, I can show it to you.

**Comment by shubham7109** [comment]
@puneet-pdx, thanks for the feedback. I have added the changes as we discussed. Ready for your review.

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt [hunk 9]
```diff
@@ -166,9 +229,11 @@ fun CalloutOptions(
                 textStyle = LocalTextStyle.current.copy(textAlign = TextAlign.End)
             )
         }
-
-        Button(onClick = viewModel::clearMapPoint) {
-            Text(text = "Clear Tap Location")
+        Spacer(modifier = Modifier.size(10.dp))
+        Button(
+            enabled = (mapPoint != null && calloutVisibility),
```
Comment (puneet-pdx) [inline microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt:234]: we should not disable this when callout is not visible. 
```suggestion
            enabled = (mapPoint != null),
```

**Review by puneet-pdx — APPROVED** [review]
@shubham7109 a minor comment. Will not hold the PR for it.

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt (hunk 9 above)
Comment (shubham7109) [reply microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt:234]: The button is to "Clear Callout map point", in my opinion having this Button enabled when Callout is not visible doesn't seem right to me. This is the same boolean condition used to display the Callout.

File: microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt (hunk 6 above)
Comment (puneet-pdx) [reply microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt]: You can revert this change, now that we have a fix for it. We don't have to do it.

Comment (shubham7109) [reply microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt:79]: Updated the change. Will rebase this PR after merge of https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/pull/482

**Comment by puneet-pdx** [comment]
@shubham7109 with these changes Animation is working for me as expected

**TapLocationScreen.kt**

```Kotlin
    // animate to a visible transition state
    val calloutVisibleState = remember { MutableTransitionState(false) }.apply {
        targetState = true
    }
    calloutVisibleState.targetState = mapPoint != null && calloutVisibility
    
    ...
    
    MapView(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues),
            arcGISMap = viewModel.arcGISMap,
            graphicsOverlays = remember { listOf(viewModel.tapLocationGraphicsOverlay) },
            onSingleTapConfirmed = viewModel::setMapPoint,
            content =
            {
                val lastMapPoint = remember { Ref<Point>() }
                lastMapPoint.value = mapPoint ?: lastMapPoint.value

                AnimatedVisibility(
                    calloutVisibleState,
                    enter = fadeIn(),
                    exit = fadeOut()
                ) {
                    lastMapPoint.value?.let {
                        Callout(
                            modifier = Modifier.wrapContentSize(),
                            location = it,
                            rotateOffsetWithGeoView = rotateOffsetWithGeoView,
                            offset = offset
                        ) {
                            Text("Tapped location:\n${it.x.roundToInt()},${it.y.roundToInt()}")
                        }
                    }
                }
            }
        )
```
https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/pull/486/files

**Comment by shubham7109** [comment]
Thanks @puneet-pdx, I've added this code to the PR https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/pull/485..

**Review by sorenoid — APPROVED** [review]

**Comment by shubham7109** [comment]
Thanks for the reviews @puneet-pdx @sorenoid. Mergering this in.

**Pull request merged**